src/server/cmd/pachd/main.go

```diff
@@ -924,7 +930,331 @@ func doFullMode(config interface{}) (retErr error) {
+	if err := dbutil.WaitUntilReady(ctx, log.StandardLogger(), env.GetDBClient()); err != nil {
+		return err
+	}
+	if err := migrations.ApplyMigrations(ctx, env.GetDBClient(), migrations.MakeEnv(nil, env.GetEtcdClient()), clusterstate.DesiredClusterState); err != nil {
+		return err
+	}
+	if err := migrations.BlockUntil(ctx, env.GetDBClient(), clusterstate.DesiredClusterState); err != nil {
+		return err
+	}
+	env.InitDexDB()
+
+	requireNoncriticalServers := !env.Config().RequireCriticalServersOnly
+
+	// Setup External Pachd GRPC Server.
+	authInterceptor := authmw.NewInterceptor(env.AuthServer)
+	loggingInterceptor := loggingmw.NewLoggingInterceptor(env.Logger())
+	externalServer, err := grpcutil.NewServer(
+		ctx,
+		true,
+		// Add an UnknownServiceHandler to catch the case where the user has a client with the wrong major version.
+		// Weirdly, GRPC seems to run the interceptor stack before the UnknownServiceHandler, so this is never called
+		// (because the version_middleware interceptor throws an error, or the auth interceptor does).
+		grpc.UnknownServiceHandler(func(srv interface{}, stream grpc.ServerStream) error {
+			method, _ := grpc.MethodFromServerStream(stream)
+			return errors.Errorf("unknown service %v", method)
+		}),
+		grpc.ChainUnaryInterceptor(
+			errorsmw.UnaryServerInterceptor,
+			version_middleware.UnaryServerInterceptor,
+			tracing.UnaryServerInterceptor(),
+			authInterceptor.InterceptUnary,
+			loggingInterceptor.UnaryServerInterceptor,
+		),
+		grpc.ChainStreamInterceptor(
+			errorsmw.StreamServerInterceptor,
+			version_middleware.StreamServerInterceptor,
+			tracing.StreamServerInterceptor(),
+			authInterceptor.InterceptStream,
+			loggingInterceptor.StreamServerInterceptor,
+		),
+	)
+
+	if err != nil {
+		return err
+	}
+	if err := logGRPCServerSetup("External Pachd", func() error {
+		txnEnv := txnenv.New()
+
+		if err := logGRPCServerSetup("Auth API", func() error {
+			authAPIServer, err := authserver.NewAuthServer(
+				authserver.EnvFromServiceEnv(env, txnEnv),
+				true, requireNoncriticalServers, true,
+			)
+			if err != nil {
+				return err
+			}
+			authclient.RegisterAPIServer(externalServer.Server, authAPIServer)
+			env.SetAuthServer(authAPIServer)
+			return nil
+		}); err != nil {
+			return err
+		}
+		if err := logGRPCServerSetup("Enterprise API", func() error {
+			e := eprsserver.EnvFromServiceEnv(env, path.Join(env.Config().EtcdPrefix, env.Config().EnterpriseEtcdPrefix), txnEnv)
+			e.IsPaused = true
+			enterpriseAPIServer, err := eprsserver.NewEnterpriseServer(
+				e,
+				true,
+			)
+			if err != nil {
+				return err
+			}
+			eprsclient.RegisterAPIServer(externalServer.Server, enterpriseAPIServer)
+			env.SetEnterpriseServer(enterpriseAPIServer)
+			// Stop workers because unpaused pachds in the process
+			// of rolling may have started them back up.
+			if err := e.StopWorkers(ctx); err != nil {
```

Comment:
should we have this call be done out of the enterprise service? Since this feature relies on restarts and startup-time logic I'm contemplating whether the logic would be easier to follow if we handle this at the enterprise service startup. WDYT?

Reply:
I’m not opposed to that, but would it live in `eprsserver.NewEnterpriseServer`, or in a new method on the enterprise server? If so, wouldn’t that new method basically be a renamed `StopWorkers`? Maybe that’s a good thing, because it abstracts the exact knowledge out a bit?